Add `Style/RedundantFreeze` to remove redudant `.freeze`

yahonda opened the pull request:
### Summary

Since Rails 6.0 will support Ruby 2.4.1 or higher `# frozen_string_literal: true` magic comment enabled by `Style/FrozenStringLiteralComment` cop is enough to make string object frozen.

### Other Information

I'd like to hear from Rails committers for this change since Rails 5.2 will supports Ruby 2.2, this change itself should not be backported but also it could make other backports complex.

matthewd commented:
I wonder how many of these are extremely performance-critical. For most, it seems fine to backport to 5.2, just to make future maintenance easier: things being *slightly* slower on 5.2 + 2.2 doesn't sound like a big deal... I'm just not sure how we'd identify any where it makes a catastrophic difference.

Otherwise, we can wait a while and then merge this in a few months, when 5.2 backports have settled down.

---

I think we might want to keep the explicit `freeze` on the tests that specifically want a frozen string, for clarity. (Maybe?)

yahonda commented:
Thanks for the review. I did not expect that backporting this change to 5-2-stable is possible.

> I think we might want to keep the explicit freeze on the tests that specifically want a frozen string, for clarity. (Maybe?)

Agreed. I have excluded 3 test files which have tests expecting a frozen string in case of backporting and respecting the intention of the tests.

jeremy commented:
Backporting to 5.2 is a nice idea. We could say Rails 5.2 supports Ruby 2.2, with this caveat, but prefers 2.3 or newer.

On the other hand, it's strange for our API to return differently mutable/frozen strings depending on the version of Ruby you're running.

matthewd commented:
> On the other hand, it's strange for our API to return differently mutable/frozen strings depending on the version of Ruby you're running.

We're already doing that on any other literals where we don't explicitly call `.freeze`, so it doesn't seem terrible... I guess there is a distinction between the performance-based freezings, and the ones where we actually care about making the value immutable (because it's being stored and reused in a constant, say).

I do think it's reasonable for us to strongly recommend a newer ruby than the oldest we support, and to bump that recommendation more freely: we want 5.2 to *work* on almost-already-EOL 2.2, but I don't see myself feeling bad it if doesn't work *well*.

yahonda commented:
Excluded two files changed by this commit. https://github.com/rails/rails/commit/8454aeeb2b8e7d1255acbaaec4e0ae3c97d55f49

yahonda commented:
Addressed conflicts. Since this pull request has been conflicting other commits
By separating commits by `rubocop -a` it is easy to rebase.

Will squash two commits when it is ready to merge.

yahonda commented:
It has been approved then I have squashed two commits.

yahonda committed:
Add `Style/RedundantFreeze` to remove redudant `.freeze`

Since Rails 6.0 will support Ruby 2.4.1 or higher
`# frozen_string_literal: true` magic comment is enough to make string object frozen.
This magic comment is enabled by `Style/FrozenStringLiteralComment` cop.

* Exclude these files not to auto correct false positive `Regexp#freeze`
 - 'actionpack/lib/action_dispatch/journey/router/utils.rb'
 - 'activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb'

It has been fixed by https://github.com/rubocop-hq/rubocop/pull/6333
Once the newer version of RuboCop released and available at Code Climate these exclude entries should be removed.

* Replace `String#freeze` with `String#-@` manually if explicit frozen string objects are required

 - 'actionpack/test/controller/test_case_test.rb'
 - 'activemodel/test/cases/type/string_test.rb'
 - 'activesupport/lib/active_support/core_ext/string/strip.rb'
 - 'activesupport/test/core_ext/string_ext_test.rb'
 - 'railties/test/generators/actions_test.rb'

yahonda commented:
As #32971 has been merged to master I wanted this pull request reviewed again.

kamipo commented on .rubocop.yml:
```diff
@@ -130,6 +130,15 @@ Style/FrozenStringLiteralComment:
 
+Style/RedundantFreeze:
+  Enabled: true
+  Exclude:
```
Why exclude these test files?

https://github.com/rails/rails/blob/master/actionpack/test/controller/test_case_test.rb
https://github.com/rails/rails/blob/master/activemodel/test/cases/type/string_test.rb
https://github.com/rails/rails/blob/master/activesupport/test/core_ext/string_ext_test.rb
https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/string/strip.rb
https://github.com/rails/rails/blob/master/railties/test/generators/actions_test.rb

matthewd replied:
In my 5 second test, Rubocop didn't seem to complain about redundant `-"foo"`. If that's the case (and deliberate?), maybe we can consider switching to that for the first two. `string_ext_test.rb` seems more like it's missing a test for the _unfrozen_ case (though I'd really prefer to see all those wrapped in an `assert_deprecated` anyway).

kamipo replied:
Yeah, I also confirmed about redundant `-"foo"` is no offence for the cop when I commented.
Since we already replaced all `String#dup` to `String#+@`, replacing redundant `String#freeze` to `String#-@` is fine to me.

yahonda replied:
Thanks for the review. I have removed these files form exclude list and replaced `String#freeze` with `String#-@` where explicit string objects are necessary.

kamipo commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -539,7 +539,7 @@ def translate_exception(exception, message)
         end
 
-        COLLATE_REGEX = /.*\"(\w+)\".*collate\s+\"(\w+)\".*/i.freeze
+        COLLATE_REGEX = /.*\"(\w+)\".*collate\s+\"(\w+)\".*/i
```
`/.*\"(\w+)\".*collate\s+\"(\w+)\".*/i` is not a string so this `.freeze` is not redundant.

yahonda replied:
Thanks for the review. Let me confirm this review first. I think you mean regular expressions is not a string. It looks like Ruby itself does not support frozen regular expressions yet.
https://bugs.ruby-lang.org/issues/8948 I'm wondering if this .freeze is actually valid or not

matthewd replied:
While it does sound like Rubocop's being over-eager in declaring it redundant, I think this instance is fine in practice. It's not a value we return anywhere, so someone would have to be trying pretty hard to even find it... and at that point a `const_set` is just as easy.

kamipo replied:
Yeah, this instance is fine in practice, and this is an only place where RuboCop's auto-correct affects in I reviewed all diff in this PR.
I suppose it is an issue for the cop, but I'm not sure whether RuboCop intentionally regard a regex literal like a string literal or not.

yahonda replied:
I have opened a RuboCop issue https://github.com/rubocop-hq/rubocop/issues/6331

koic replied:
Thank you for opening the RuboCop issue. I think this is a false positive of RuboCop. I opened a PR rubocop-hq/rubocop#6333 to fix, but it may take some time to the next new release of RuboCop (maybe 0.59.3 or 0.60.0) and subsequent Code Climate support. Until then it will depend on hand craft.

kamipo commented on .rubocop.yml:
```diff
@@ -130,6 +130,15 @@ Style/FrozenStringLiteralComment:
+    - 'actionpack/test/controller/test_case_test.rb'
+    - 'activemodel/test/cases/type/string_test.rb'
+    - 'activesupport/test/core_ext/string_ext_test.rb'
+    - 'activesupport/lib/active_support/core_ext/string/strip.rb'
```
This is not redundant freeze.
https://github.com/rails/rails/blob/d03cc170c1bdad15245be7f2681a265fd2bd0d5d/activesupport/lib/active_support/core_ext/string/strip.rb#L24

kamipo replied:
This `"".freeze` is obviously redundant, there is no reason to exclude it.
https://github.com/rails/rails/blob/d03cc170c1bdad15245be7f2681a265fd2bd0d5d/activesupport/lib/active_support/core_ext/string/strip.rb#L23

yahonda commented on .rubocop.yml:
```diff
@@ -130,6 +130,11 @@ Style/FrozenStringLiteralComment:
+Style/RedundantFreeze:
+  Enabled: true
+  Exclude:
+    - 'activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb'
```
Just excluded this file to avoid false positive until the fix made by https://github.com/rubocop-hq/rubocop/pull/6333 is available as the newer version of RuboCop, 0.60 maybe and Code Climate supports the new version of RuboCop to avoid auto correct by the false positive.

kamipo commented on actionpack/lib/action_dispatch/journey/router/utils.rb:
```diff
@@ -18,33 +18,33 @@ def self.normalize_path(path)
-          PATH     = /[^#{UNRESERVED}#{SUB_DELIMS}:@\/]/.freeze
+          FRAGMENT = /[^#{UNRESERVED}#{SUB_DELIMS}:@\/\?]/
+          SEGMENT  = /[^#{UNRESERVED}#{SUB_DELIMS}:@]/
+          PATH     = /[^#{UNRESERVED}#{SUB_DELIMS}:@\/]/
```
I re-searched by "/.freeze" in the diff and found extra affected regexs.

yahonda replied:
Thank you for finding other false positives. Addressed and exclude 'actionpack/lib/action_dispatch/journey/router/utils.rb' also.
